nmigen/lib/cdc.py

```diff
@@ -95,6 +95,61 @@ def elaborate(self, platform):
         return m
 
 
+class AsyncFFSynchronizer(Elaboratable):
+    """Synchronize deassertion of an asynchronous signal.
+
+    The signal driven by the :class:`AsyncFFSynchronizer` is asserted asynchronously and deasserted
+    synchronously, eliminating metastability during deassertion.
+
+    This synchronizer is primarily useful for resets and reset-like signals.
+
+    Parameters
+    ----------
+    i : Signal(1), in
+        Asynchronous input signal, to be synchronized.
+    o : Signal(1), out
+        Synchronously released output signal.
+    domain : str
+        Name of clock domain to reset.
+    stages : int, >=2
+        Number of synchronization stages between input and output. The lowest safe number is 2,
+        with higher numbers reducing MTBF further, at the cost of increased deassertion latency.
+    """
+    def __init__(self, i, o, *, domain="sync", stages=2, async_edge="pos"):
+        _check_stages(stages)
+
+        self.i = i
+        self.o = o
+
+        self._domain = domain
+        self._stages = stages
+
+        if async_edge not in ("pos", "neg"):
+            raise ValueError("AsyncFFSynchronizer async edge must be one of 'pos' or 'neg', not {!r}"
+                             .format(async_edge))
+        self._edge = async_edge
+
+    def elaborate(self, platform):
+        m = Module()
```

Comment:
Do we need the ability to swap in a platform-specific module here?

Reply:
Yes, absolutely. In fact we *only* need it here and not in `ResetSynchronizer`; the reason platform overrides exist is to make sure CDC constraints are correct, and `ResetSynchronizer` itself doesn't do any CDC anymore.

Reply:
Ah, yes. Hm... It doesn't really want to be called `get_reset_sync` anymore then does it? So much for not touching the platform code.  `get_async_ff_sync`?

Reply:
> It doesn't really want to be called `get_reset_sync` anymore then does it? `get_async_ff_sync`?

Sounds good to me.

> So much for not touching the platform code.

That was inevitable because of the dual edge support I think. (We can postpone that; I do have every supported toolchain installed though.)

Reply:
Mm this also means `_max_input_delay` needs to move to `AsyncFFSynchronizer`.

> That was inevitable because of the dual edge support

Should I add an edge parameter to `ResetSynchronizer` while I'm  in here?

Reply:
After reviewing the current code I suggest postponing platform support for configurable edge. This is because only Intel and Xilinx override CDC currently, and the three Xilinx platforms would really benefit from having clock domains with negative polarity resets for implementing this. You can insert an assertion in every platform override that the polarity is what it expects.

Reply:
> Mm this also means `_max_input_delay` needs to move to `AsyncFFSynchronizer`.

Yes.

> Should I add an edge parameter to `ResetSynchronizer` while I'm in here?

No, not yet. I'm actually not sure yet how configurable polarity resets should work, given that `ResetSignal` is late-bound. Not being able to write something like `m.If(ResetSignal()):` to check if you are being reset, and having to branch on polarity, would be really awful. Having `cd.rst` always have the logical polarity is confusing when you're driving it. Inverting the polarity of `ResetSignal()` that is being driven on LHS is technically complex. It's a can of worms... unfortunately one that is required for ASICs.